src/bin/pg_autoctl/supervisor.c

```diff
@@ -724,8 +651,77 @@ supervisor_restart_service(Supervisor *supervisor, Service *service, int status)
 		return false;
 	}
 
-	/* we could restart, update our service start time */
-	service->startTime = now;
+	return true;
+}
+
+
+/*
+ * supervisor_count_restarts returns true when we have restarted more than
+ * SUPERVISOR_SERVICE_MAX_RETRY in the last SUPERVISOR_SERVICE_MAX_TIME period
+ * of time.
+ */
+static bool
+supervisor_may_restart(Service *service)
```

Comment:
I think the code to could be a bit simpler (no loop necessary afaict). See pseudo code below:
Initially: 
```c
next_position_to_insert = 0
startTime = empty array of size 5 (not 6 like now)
count = 0 // number of entries in the ringbuffer 
```
Every time a restart happens
```c
// If this is not the case 5 restarts have not yet happend so we can try again for sure.
if (count == 5) {
    
    if (now - starTime[next_position_to_insert] <= 15) {
         
        return false;
    }
}

starTime[next_position_to_insert] = now
if (count < 5) {
    count += 1
}
next_position_to_insert = (next_position_to_insert + 1) % 5
return true
```

Reply:
Done now, thanks for your input. I think we're back to the previous coding, but with all the intermediate restart time stamps rather than some smarts about keeping only one of them and a weird third variable and reset mechanism. Overall I like the new code better, thanks!